### ISSUE TITLE
docdatetime attribute value is not consistent with documentation

### ISSUE THREAD

**Issue by thom4parisot** [asciidoctor/asciidoctor.js#517]
Documentation states the following:

> Attributes that store dates and times (e.g., docdatetime) are always formatted like 2019-01-04 19:26:06 GMT+0000.

Although it happens that `node.getDocument().getAttribute('docdatetime')` prints `2018-07-09 10:37:11 Central European Summer Time`.

Is it the expected output of the timezone?

**Comment by Mogztter** [comment]
That's a good question :smile: 

I know that we had some issues to coerce Ruby Dates to JavaScript Dates in Opal so maybe the JavaScript implementation is not 100% compliant or maybe it's a bug or maybe the documentation is outdated.
Anyway, I will add some tests to understand what's going on!

**Comment by Mogztter** [comment]
I do have the same behavior with Asciidoctor:

**Ruby API**
```
$ cat test.rb 
```
```rb
require 'asciidoctor'

doc = Asciidoctor.load('test')
p doc.attr('docdatetime')
```

```
$ ruby test.rb
2018-07-09 17:40:37 CEST
```

**Ruby CLI**
```
$ asciidoctor -v                
Asciidoctor 1.5.7.1 [https://asciidoctor.org]
Runtime Environment (ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-linux]) (lc:UTF-8 fs:UTF-8 in:UTF-8 ex:UTF-8)
```
```
$ asciidoctor test.adoc
```
```  
$ cat test.html | grep "Time is"
<p>Time is 2018-07-09 17:41:41 CEST</p>
```

@mojavelinux Is it a localization/timezone bug or an outdated documentation ?

**Comment by Mogztter** [comment]
https://github.com/asciidoctor/asciidoctor/blob/217dc6fb3ec09ad522d86bb82ba80ec904486e27/lib/asciidoctor/document.rb#L499-L503

So we are using `strftime` with `%T %Z`: 

> %T - 24-hour time (%H:%M:%S)
> %Z - Abbreviated time zone name or similar information.  (OS dependent)

https://ruby-doc.org/core-2.2.0/Time.html#method-i-strftime

But interestingly if it fails, we are using `%z` which is the format described in the documentation: 
> %z - Time zone as hour and minute offset from UTC (e.g. +0900)

I would expect `doctime` and `docdatetime` to be UTC and `localdatetime` and `localtime` to be local time (with timezone).
It can make sense to use the time zone name if the value is for display only but I don't think it's the case here. So I would advocate for a numeric (not OS dependent) date time.

If you are using the Asciidoctor Ruby CLI you can use `SOURCE_DATE_EPOCH` to use an UTC date:

```
$ SOURCE_DATE_EPOCH=1531151244 asciidoctor test.adoc
```
```
$ cat test.html | grep "Time is"                    
<p>Time is 2018-07-09 15:47:24 UTC</p>
```

But it will probably not work on Asciidoctor.js because the implementation to resolve environment variables is missing (in Opal).

**Comment by iliabylich** [comment]
Opal can handle ENV variables (by dumping all ENV variables to the file, so these values has effect only on compilation):
```
$ A=1 opal -rnodejs -e 'p ENV["A"]'
"1"
```

But TZ (and Time-related) stuff is too complex to make it working properly in JS (in all browsers). 
``` js
> new Date().$strftime("%Z")
"Москва, стандартное время"
> new Date().$strftime("%z")
"+0300"
```
Looks like it's similar to the original issue.

Modern browsers have a new feature called "internationalization API" that handles it very well:
```
> new Intl.DateTimeFormat('en-US', { hour: 'numeric', timeZoneName: 'short', timeZone: 'UTC' }).format(new Date())
"4 PM UTC"
> new Intl.DateTimeFormat('en-US', { hour: 'numeric', timeZoneName: 'short' }).format(new Date())
"7 PM GMT+3"
```

It even supports converting TZ and complex formatting (so I think it's as powerful as `Time#strftime` in Ruby), and it's implemented in Node.

From what I remember it was discussed many times in the Opal repo, but we rejected it because of the phantomjs, so maybe it's time to revisit it.

**Comment by Mogztter** [comment]
> Opal can handle ENV variables (by dumping all ENV variables to the file, so these values has effect only on compilation):

Thanks for your input.
Opal could potentially handle environment variables at runtime in Node.js only ?

> Modern browsers have a new feature called "internationalization API" that handles it very well:
It even supports converting TZ and complex formatting (so I think it's as powerful as Time#strftime in Ruby), and it's implemented in Node.
From what I remember it was discussed many times in the Opal repo, but we rejected it because of the phantomjs, so maybe it's time to revisit it.

Indeed! Aa far as I remember we dropped PhantomJS a few months ago so definitely something to revisit :+1:

**Comment by oncletom** [comment]
What's nice about the Internationalisation API is it's available in Node as well 👍 (it's part of a sub-spec of ECMAScript).

🔎 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat

**Comment by mojavelinux** [comment]
See https://github.com/asciidoctor/asciidoctor/issues/2770. In the discussion for that issue, I learned that Windows likes to use elaborate timezone names instead of the standard codes that are recognized internationally. In fact, Ruby acknowledges that the result of %Z is OS dependent. :upside_down_face: As a result, we get all kinds of crazy problems.

I'm normalizing this behavior in 1.5.8 by using the timezone offset instead (%z). That sidesteps the encoding issue, but also makes the output more portable.

It's important to keep in mind that the date/time attributes in Asciidoctor are currently formatted strings. We could discuss changing that to something more standard, but that would have to happen in core (and would be a deviation from AsciiDoc Py compliance).

**Comment by Mogztter** [comment]
In Asciidoctor.js 1.5.9, here's the output:

```js
const asciidoctor = require('asciidoctor.js')()

const doc = asciidoctor.load('test')
console.log(doc.getAttribute('docdatetime'))
```

```
2019-02-05 09:46:37 +0100
```


> Attributes that store dates and times (e.g., docdatetime) are always formatted like 2019-01-04 19:26:06 GMT+0000.

Since 1.5.8, dates and times are actually formatted like `2019-01-04 19:26:06 +0100` or `2019-01-04 19:26:06 UTC` (if the offset is 0). So we should update the documentation: https://asciidoctor.org/docs/user-manual/#language-support

**Comment by mojavelinux** [comment]
Correct. We stopped trying to use the fancy timezone because it varies widely by operating system and region and causes all kinds of encoding problems.
